genrl/deep/common/trainer.py

```diff
@@ -1,516 +1,472 @@
-        """
-        if self.load_model is not None:
-            self.load()
-
-        for epoch in range(self.epochs):
-            self.agent.epoch_reward = np.zeros(self.env.n_envs)
-
-            self.agent.rollout.reset()
-
-            state = self.env.reset()
-            values, done = self.agent.collect_rollouts(state)
-
-            self.agent.get_traj_loss(values, done)
-
-            self.agent.update_policy()
-
-            if epoch % self.log_interval == 0:
-                self.logger.write(
-                    {
-                        "Timestep": epoch * self.agent.rollout_size,
-                        "Episode": epoch,
-                        **self.agent.get_logging_params(),
-                    }
-                )
-
-            if self.render:
-                self.env.render()
-
-            if self.save_interval != 0 and epoch % self.save_interval == 0:
-                self.save(epoch * self.agent.batch_size)
-
-        self.env.close()
-        self.logger.close()
+from abc import ABC
+from typing import Any, List, Optional, Type, Union
+
+import gym
+import numpy as np
+import torch
+
+from ...environments import VecEnv
+from .buffers import PrioritizedBuffer, ReplayBuffer
+from .logger import Logger
+from .utils import save_params, set_seeds
+
+
+class Trainer(ABC):
+    """
+    Base Trainer class. To be inherited specific usecases.
+
+    :param agent: Algorithm object
+    :param env: Environment
+    :param logger: Logger object
+    :param x_axis: Key plotted on x_axis 
+    :param buffer: Buffer Object
+    :param off_policy: Is the algorithm off-policy?
+    :param save_interval: Model to save in each of these many timesteps
+    :param render: Should the Environment render
+    :param max_ep_len: Max Episode Length
+    :param distributed: (True if distributed training is enabled, else
+False (To be implemented))
+    :param ckpt_log_name: Model checkpoint name
+    :param steps_per_epochs: Steps to take per epoch?
+    :param epochs: Total Epochs to train for
+    :param device: Device to train model on
+    :param log_interval: Log important params every these many steps
+    :param batch_size: Size of batch
+    :param seed: Set seed for reproducibility
+    :param deterministic_actions: Take deterministic actions during training.
+    :type agent: object
+    :type env: object
+    :type logger: object
+    :type x_axis: str
+    :type buffer: object
+    :type off_policy: bool
+    :type save_interval: int
+    :type render: bool
+    :type max_ep_len: int
+    :type distributed: bool
+    :type ckpt_log_name: string
+    :type steps_per_epochs: int
+    :type epochs: int
+    :type device: string
+    :type log_interval: int
+    :type batch_size: int
+    :type seed: int
+    :type deterministic_actions: bool
+    """
+
+    def __init__(
+        self,
+        agent: Any,
+        env: Union[gym.Env, VecEnv],
+        log_mode: List[str] = ["stdout"],
+        x_axis: str = "timestep",
```

Comment:
Rename the variable from `x_axis` to something more understandable like `log_key` perhaps.